### PR TITLE
Enable managed memory feature

### PR DESCRIPTION
**Pull request by Watson1978**
This patch aims to goal to get Ruby's GC to work well.

For example, Ruby indicates Magick::Image object is only 40 bytes.

```ruby
require 'rmagick'
require 'objspace'

img = Magick::Image.new(400, 400)
puts ObjectSpace.memsize_of(img) #=> 40 bytes
```

However, ImageMagick allocates a lot of memory and associates to Magick::Image object.
And Ruby doesn't know that.

Please execute the following code. 

```ruby
4000.times do
  img = Magick::Image.new(400, 400)
end
```

Ruby will recognize that 40 * 4000 = 156 KB of memory has been allocated.
It may be low value as a threshold to run GC mark/sweep.

However, in fact, RMagick has been allocating more memory (156 KB + ImageMagick allocated memory).

When Ruby knows the correct memory usage, the GC works properly, and it makes system working more stably with other gems.

To solve this problem, I think we can select a way from

1. Use Ruby allocating function (xmalloc/xrealloc/xfree) in ImageMagick.
2. Notify ImageMagick allocated memory size using `rb_gc_adjust_memory_usage()` Ruby API.

I think `1.` is easy way as solution.

### Before/After
memory_usage

### Test code
```ruby
1000.times do |i|
  img = Magick::Image.new(1000, 1000)
end
```

### How to disable managed memory
Set some value to `RMAGICK_DISABLE_MANAGED_MEMORY` environment variable at command line

```
$ RMAGICK_DISABLE_MANAGED_MEMORY=true ruby test.rb
```

Or set some value to `ENV['RMAGICK_DISABLE_MANAGED_MEMORY']` before `require 'rmagick'` line.

```ruby
ENV['RMAGICK_DISABLE_MANAGED_MEMORY'] = 'true'
require 'rmagick'
```

### CODE REVIEW

**Comment by mockdeep** [comment]
Wow, this is pretty awesome! It looks like it was also mostly < 1GB.

**Comment by Watson1978** [comment]
I update PR description and found no difference by new test code.
However, I found implementation miss in ImageMagick.

When I fix the miss in ImageMagick and uses this PR, it brings huge improvement for memory usage.
memory_usage2

**Comment by Watson1978** [comment]
@dlemstra I sent PR to ImageMagick6 at https://github.com/ImageMagick/ImageMagick6/pull/49
Can you review it ?
And if it has no problem, can you port it to ImageMagick 7 ?

**Comment by mockdeep** [comment]
Holy cow! That's a huge difference!

**Comment by Watson1978** [comment]
Seems that my patch for ImageMagick has shipped with [ver. 6.9.10-49](https://github.com/ImageMagick/ImageMagick6/commit/a7ae3f482ce47569a6b8d5a6d3b4597ad286593a) and [ver. 7.0.8-49](https://github.com/ImageMagick/ImageMagick/commit/283cf06bec52ff418a68d640176901207c7c6980)

**Comment by Watson1978** [comment]
I think this PR ready to merge to master. 

@mockdeep @dlemstra  What do you think about ?

File: ext/RMagick/rmmain.c [hunk 1]
```diff
@@ -123,20 +103,35 @@ static void rm_free(void *ptr)
...
+
+    if (disable)
     {
-        rb_define_const(Module_Magick, "MANAGED_MEMORY", Qfalse);
+        managed_memory_enable(Qfalse);
+        return;
     }
+
+#if defined(_WIN32)
+#if defined(IMAGEMAGICK_GREATER_THAN_EQUAL_6_9_0)
```
Comment (dlemstra) [inline ext/RMagick/rmmain.c]: Can you clarify why we don't enable it for versions lower than 6.9.0 on Windows?

Comment (Watson1978) [reply ext/RMagick/rmmain.c]: I tried various things, but it crashes silently when it is old IM for some reason.

https://ci.appveyor.com/project/mockdeep/rmagick/builds/24706171

I gave up to investigate it... sorry😖

Comment (dlemstra) [reply ext/RMagick/rmmain.c]: No problem was just wondering why. Could you add this as a comment in the code or in your squash commit message?

File: benchmarks/memory/rmagick.gnuplot [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+set grid ytics linewidth 1 linetype -1
+
+set style line 1 lt 1 lc rgbcolor "#3465a4" lw 2.5 pt 7 ps 1
+set style line 2 lt 1 lc rgbcolor "#ff6347" lw 2.5 pt 7 ps 1
+set style line 3 lt 1 lc rgbcolor "#888a85" lw 2.5 pt 5 ps 1
+
+set xlabel "Nth loop"
+set ylabel "Memory usage (MiB)"
+
+plot "before.csv" with line linestyle 1, "after.csv" with line linestyle 2
```
Comment (mockdeep) [inline benchmarks/memory/rmagick.gnuplot]: Can you share how this file is used? Should we have a `rake` task to automate it somehow? Or maybe we don't need it?

**Comment by mockdeep** [comment]
Great work on this @Watson1978! This looks like it's going to have a *huge* impact on performance.

File: benchmarks/memory/rmagick.gnuplot (hunk 2 above)
Comment (Watson1978) [reply benchmarks/memory/rmagick.gnuplot]: 1. Install `gnuplot`.
2. Retrieve two performance data to draw graph.
```
# Retrieve data before applying patches.
$ rake build
$ gem install pkg/rmagick-3.1.0.gem
$ cd benchmarks/memory
$ ruby image_new.rb > before.csv


# Apply the PR then retrieve improved data
$ rake build
$ gem install pkg/rmagick-3.1.0.gem
$ cd benchmarks/memory
$ ruby image_new.rb > after.csv
```

3. Launch `gnuplot` and execute `load 'rmagick.gnuplot'` command in prompt then the performance graph will be drew.
```
$ /usr/local/bin/gnuplot

	G N U P L O T
	Version 5.2 patchlevel 7    last modified 2019-05-29

	Copyright (C) 1986-1993, 1998, 2004, 2007-2018
	Thomas Williams, Colin Kelley and many others

	gnuplot home:     http://www.gnuplot.info
	faq, bugs, etc:   type "help FAQ"
	immediate help:   type "help"  (plot window: hit 'h')

Terminal type is now 'qt'
gnuplot> load 'rmagick.gnuplot'
```

File: ext/RMagick/rmmain.c (hunk 1 above)
Comment (Watson1978) [reply ext/RMagick/rmmain.c]: Added comment at https://github.com/rmagick/rmagick/pull/697/commits/ec3d796da2d8e530369571c51e9aba76cd8a1520

File: benchmarks/memory/rmagick.gnuplot (hunk 2 above)
Comment (Watson1978) [reply benchmarks/memory/rmagick.gnuplot]: Added README how to draw the performance graph.

Comment (mockdeep) [reply benchmarks/memory/rmagick.gnuplot]: Excellent, thanks!

**Pull request merged**